### ISSUE TITLE
Questions for Converting From Provider

### ISSUE THREAD

**Issue by yasinarik** [jonataslaw/getx#525]
This is not a bug issue but since GetX is fairly new, I wanted to help its development by asking some questions. 

As an entrepreneur I do intensive coding as well as marketing. So I’d like to share some opinions and point some areas under the shade to create a collective help which may help you and GetX to reach more people especially teams and corporate firms that develop complex apps. 

Before starting, I’d like the say that I’m very glad that the community has people like you. I appreciate the work done for GetX. Thank you. 

1) I’m using Provider with ChangeNotifier for an highly complex app. That many Provider classes cause a performance problem. (It is not seen but eventually it happens as shown in debugging tools.)

Also using providers is difficult because of syntax and the need of context. 

The reactive side of the GetX, the Obx, may also cause performance issues since they are basically open streams.

So I figured out that the best way for a state management is using the update() side of the GetX. That way, updates occur only on demand. The syntax is easy, no context and no calling the controller class inside the main build method of a widget class. Also I can choose the exact moment to execute update(); function which gives me more control. Immediate reactiveness is not what I always want. 

2) When using GetBuilder, what if we need multiple controller classes? In Provider package we have the MultiProvider package. It’s useful to wrap other widgets. GetX does need a wrap too. 

How can I depend on multiple GetX controllers inside the GetBuilder? (Maybe wrapping them over and over just works but is there a better syntax for readability?)

3) Provider package uses the closest provider class declared in the widget tree. I have post and profile screens in my app. (You can think it like Instagram where you can constantly navigate to a post than its user profile than another post than a tagged person’s profile over and over.)

I use some of the provider classes as the logic controller instead of doing logic stuff inside the StatefulWidget. 

For example there is a ProfileScreenUI class which holds UI related to a user profile. There is also a ProfileScreenProvider class which is consumed by each of the ProfileScreens. 

A user can navigate to another user’s profile page over and over and do stuff which changes their state. 

UserA—>UserC—>UserB—> on it goes on. The screens are totally unique to the corresponding users. So each time the user navigates to a new user (or even the same user’s profile), I create another ProfileScreenUI class using a separate instance of ProfileScreenProvider class as a controller. 

Because that I give each of the new ProfileScreenUI classes a provider class, they always use the closest one up in the widget tree which is the new ProfileScreenProvider instance. 

Can I do this with GetX?

**Comment by roipeker** [comment]
@yasinarik If you have any questions on how to use the package, please use the proper channels:
Look for **GetX Community channels** in the [readme](https://github.com/jonataslaw/getx), Telegram and Discord are very responsive, we'll be happy to help you out.
@jonataslaw / @Nipodemos  would you mind closing the issue?

**Comment by jonataslaw** [comment]
When using GetBuilder, what if we need multiple controller classes? In Provider package we have the MultiProvider package. It’s useful to wrap other widgets. GetX does need a wrap too.
**- Bindings**

How can I depend on multiple GetX controllers inside the GetBuilder? (Maybe wrapping them over and over just works but is there a better syntax for readability?
Get.find<OtherController>()

Sorry, but the other questions are supportive.
This channel is exclusive for opening issues, or requesting resources.
The platform channels are suitable for you to get support from the community.

Since this is not a description of bugs or a feature request, I am closing this.

**Comment by Nipodemos** [comment]
> How can I depend on multiple GetX controllers inside the GetBuilder? (Maybe wrapping them over and over just works but is there a better syntax for readability?
> **Get.find()**

it does not work jonatas
let's say i have a text that half from one controller and half of another one

if i don't call the update method of the controller that was in the GetBuilder, it won't rerender: here is an example:
![hdVeMnjVew](https://user-images.githubusercontent.com/11494727/92010899-b9ca4a00-ed20-11ea-9671-94e74676dce0.gif)
here is reproducible code:

```dart
class FirstController extends GetxController {
  String firstWord = "Hey";

  void updateWord() {
    this.firstWord = "Hello";
    print('updating first part to "Hello"');
    update();
  }
}

class SecondController extends GetxController {
  String secondWord = "World";

  updateWord() {
    this.secondWord = "to you!";
    print("updating second part to 'to you!'");
    update();
  }
}

class GetBuilderViewPage extends StatelessWidget {
  final secondController = Get.put<SecondController>(SecondController());
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text('GetBuilderViewPage')),
      body: Container(
        child: Column(
          children: [
            GetBuilder<FirstController>(
              init: FirstController(),
              builder: (controller) {
                return Text(
                  '${controller.firstWord} ${secondController.secondWord}',
                  style: TextStyle(fontSize: 24),
                );
              },
            ),
            RaisedButton(
              child: Text('update to "Hello to you!"'),
              onPressed: () {
                // if i don't call the update method of the first controller, it will not rerender
                // Get.find<FirstController>().updateWord();
                Get.find<SecondController>().updateWord();
              },
            )
          ],
        ),
      ),
    );
  }
}
```

**Comment by jonataslaw** [comment]
As I recall (I haven't used Provider for over a year), multiproviders are just for injecting multiple ChangeNotifier classes into the tree.
Bindings are a more readable and decoupled way of doing this.

When it comes to nesting GetBuilder, I believe that this is impossible in any type of state management other than Getx (reactive), and perhaps Mobx, but Provider, Bloc, nor Simple State Manager of Get will be able to reproduce this example successfully, simply because they are widgets connected to controllers/blocs/changeNotifiers.

There is Value + SimpleBuilder that I didn't even document that is the most effective way to do this, however, it is not based on Streams, so it loses a little bit of power, however, it has the build time 6x faster than Rx , Bloc, Provider, or even a StreamController + StreamBuilder.

I'm testing this to see if I can create a notification pool, so that I can add Streams, using the simplicity of the hashset notifier.

**Comment by jonataslaw** [comment]
Try:
```dart

class FirstController extends GetxController {
  final firstWord = Value<String>("Hey");

  void updateWord() {
    this.firstWord.value = "Hello";
    print('updating first part to "Hello"');
  }
}

class SecondController extends GetxController {
  final secondWord = Value<String>("World");

  updateWord() {
    this.secondWord.value = "to you!";
    print("updating second part to 'to you!'");
  }
}

class GetBuilderViewPage extends StatelessWidget {
  final controller = Get.put<FirstController>(FirstController());
  final secondController = Get.put<SecondController>(SecondController());
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text('GetBuilderViewPage')),
      body: Container(
        child: Column(
          children: [
            SimpleBuilder(
              builder: (_) {
                return Text(
                  '${controller.firstWord.value} ${secondController.secondWord.value}',
                  style: TextStyle(fontSize: 24),
                );
              },
            ),
            RaisedButton(
              child: Text('update to "Hello to you!"'),
              onPressed: () {
                // if i don't call the update method of the first controller, it will not rerender
                Get.find<FirstController>().updateWord();
                Get.find<SecondController>().updateWord();
              },
            )
          ],
        ),
      ),
    );
  }
}

```

**Comment by jonataslaw** [comment]
I need an opinion on the Value.
It has about 100 lines of code, it doesn't work with Streams, it's fast, but it's not as powerful as Rx.
I don't know if I present it as an intermediate alternative (it will be confusing 3 State Managers in 1 package).
If I try to migrate it with Rx (I don't know if it would be a good option, Workers and bindStream are basically Rx's biggest competitive advantage over other State managers).
I don't know what I do.

**Comment by Nipodemos** [comment]
> ```dart
>  final firstWord = Value<String>("Hey");
> ```
man you always have something under your sleeves huh? 🤯

**Comment by jonataslaw** [comment]
I always want to have the best performance option, the best build time option, and the lowest memory consumption, and sometimes I insert these experimental things to test benchmarks.
Value is undoubtedly the lightest and simplest way to manage states, if we used dart extensions, to create something like:

final name = "jonny" .reactive;

It would be so much easier than anything else.

But I don't know how users would react to the addition of yet another intermediate State manager.

**Comment by Nipodemos** [comment]
that's a hard question to answer
to be honest it really would not be so nice to have another intermediate

if it was possible for a widget that listen to every `update()` that is called in controllers without change the variable declaration 
( like a SimpleBuilder but without have to type `text.value` and `Value<string>('')` it would be perfect
but i can imagine that is not that simple

**Comment by jonataslaw** [comment]
Unfortunately this is not possible with dart, because it does not let you extend String, int, num, and there is no way to override the setter operator "=". So that will unfortunately not happen, maybe never.

**Comment by Nipodemos** [comment]
well, getbuilder is not made to help in every situation anyway, it is made to simple things

maybe then we should add to docs the limitations of getBuilder instead of trying to make it get more and more closer to getx
